tsl/src/chunk.c

```diff
@@ -507,3 +508,188 @@ chunk_drop_replica(PG_FUNCTION_ARGS)
+	return id_list;
+}
+
+static List *
+chunk_id_list_exclusive_right_merge_join(const List *an_list, const List *dn_list)
+{
+	/*
+	 * merge join two sorted list and return only values which exclusively
+	 * exists in the right target (dn_list list)
+	 */
+	List *result = NIL;
+	const ListCell *a = list_head(an_list);
+	const ListCell *b = list_head(dn_list);
+	for (;;)
+	{
+		if (a && b)
+		{
+			int compare = list_int_cmp_compat(a, b);
+			if (compare == 0)
+			{
+				/* a = b */
+				a = lnext_compat(an_list, a);
+				b = lnext_compat(dn_list, b);
+			}
+			else if (compare < 0)
+			{
+				/* min a */
+				/* chunk exists only on the access node */
+				a = lnext_compat(an_list, a);
+			}
+			else
+			{
+				/* min b */
+				/* chunk exists only on the data node */
+				result = lappend_int(result, lfirst_int(b));
+				b = lnext_compat(dn_list, b);
+			}
+		}
+		else if (a)
+		{
+			/* chunk exists only on the access node */
+			a = lnext_compat(an_list, a);
+		}
+		else if (b)
+		{
+			/* chunk exists only on the data node */
+			result = lappend_int(result, lfirst_int(b));
+			b = lnext_compat(dn_list, b);
+		}
+		else
+		{
+			break;
+		}
+	}
+	return result;
+}
+
+/*
+ * chunk_drop_stale_chunks:
+ *
+ * This function drops chunks on a specified data node if those chunks are
+ * not known by the access node (chunks array).
+ *
+ * This function is intended to be used on the access node and data node.
+ */
+Datum
+chunk_drop_stale_chunks(PG_FUNCTION_ARGS)
+{
+	char *node_name = PG_ARGISNULL(0) ? NULL : NameStr(*PG_GETARG_NAME(0));
+	ArrayType *chunks_array = PG_ARGISNULL(1) ? NULL : PG_GETARG_ARRAYTYPE_P(1);
+	DistUtilMembershipStatus membership;
+
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+
+	/* execute according to the node membership */
+	membership = dist_util_membership();
+	if (membership == DIST_MEMBER_ACCESS_NODE)
+	{
+		StringInfo cmd = makeStringInfo();
+		bool first = true;
+
+		if (node_name == NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("node_name argument cannot be NULL")));
+		if (chunks_array != NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("chunks argument cannot be used on the access node")));
+
+		/* get an exclusive lock on the chunks catalog table to prevent new chunk
+		 * creation during this operation */
+		LockRelationOid(ts_catalog_get()->tables[CHUNK].id, AccessExclusiveLock);
+
+		/* generate query to execute drop_stale_chunks() on the data node */
+		appendStringInfo(cmd, "SELECT _timescaledb_internal.drop_stale_chunks(NULL, array[");
+
+		/* scan for chunks that reference the given data node */
+		ScanIterator it = ts_chunk_data_nodes_scan_iterator_create(CurrentMemoryContext);
+		ts_chunk_data_nodes_scan_iterator_set_node_name(&it, node_name);
+		ts_scanner_foreach(&it)
+		{
+			TupleTableSlot *slot = ts_scan_iterator_slot(&it);
+			bool PG_USED_FOR_ASSERTS_ONLY isnull = false;
+			int32_t node_chunk_id;
+
+			node_chunk_id =
```

Comment:
@pmwkaa I don't think that the chunk_id is the SAME on the AN and the DN. Looks like we might have to send the chunk_names instead of ids..

Reply:
We are iterating over `chunk_data_node` table in order to get chunk id on the data node. This is a mapping table for chunks we use on the access node: https://github.com/timescale/timescaledb/blob/main/src/ts_catalog/catalog.h#L591. S

Reply:
Ah, ok. Cool